### DISCUSSION TITLE
prepareVectorTag and accumulateTaggedLocalResidual

### DISCUSSION THREAD

**Discussion by dingqiushi** [idaholab/moose#21386] [Q&A General] [answered]
My MultiKernel.cpp inherited from the existing Kernel.cpp in moose, and I wanted to add multi-physics calculations, so I overloaded the computeResidual() function with an extra layer of p-loops, but this will cause some problems when opening the displacement (even if the displacement field was 0).
```
void Kernel::computeResidual()
{
prepareVectorTag(_assembly, _var.number());
precalculateResidual();
for (_i = 0; _i < _test.size(); _i++)
  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
    _local_re(_i) += _JxW[_qp] * _coord[_qp] * computeQpResidual();
accumulateTaggedLocalResidual();
if (_has_save_in)
{
   Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
   for (const auto & var : _save_in)
     var->sys().solution().add_vector(_local_re, var->dofIndices());
}
}

void MultiKernel::computeResidual()
{
	for (unsigned int p = 0; p < _n_equation; ++p)
	{
		prepareVectorTag(_assembly, p);
		for (_qp = 0; _qp < _qrule->n_points(); _qp++)
		{
			precalculateResidual();
			for (_i = 0; _i < _test.size(); _i++)
				_local_re(_i)+=_JxW[_qp] * _coord[_qp] * computeQpResidual(p);
			

		}
		accumulateTaggedLocalResidual();
	}
}
```
I tried to change the way I wrote it so that I could ensure that the calculations for 0 displacement field and no displacement field were exactly the same, but obviously neither was the result I wanted.
```
void MultiKernel::computeResidual()
{
	for (unsigned int p = 0; p < _n_equation; ++p)
	{
		prepareVectorTag(_assembly, _var.number());
		for (_qp = 0; _qp < _qrule->n_points(); _qp++)
		{
			precalculateResidual();
			for (_i = 0; _i < _test.size(); _i++)
				_local_re(_i)+=_JxW[_qp] * _coord[_qp] * computeQpResidual(p);
			

		}
		accumulateTaggedLocalResidual();
	}
}
```
I would like to know the exact meaning and use of the functions prepareVectorTag() and accumulateTaggedLocalResidual(), and my intuition is that there is a problem with placing these two functions in a p-loop. What is the problem with putting them in the p-loop I defined?
```
void TaggingInterface::prepareVectorTag(Assembly & assembly, unsigned int ivar)
{
  _re_blocks.resize(_vector_tags.size());
  mooseAssert(_vector_tags.size() >= 1, "we need at least one active tag");
  auto vector_tag = _vector_tags.begin();
  for (MooseIndex(_vector_tags) i = 0; i < _vector_tags.size(); i++, ++vector_tag)
    _re_blocks[i] = &assembly.residualBlock(ivar, *vector_tag);
  _local_re.resize(_re_blocks[0]->size());
}

void TaggingInterface::accumulateTaggedLocalResidual()
{
  for (auto & re : _re_blocks)
    *re += _local_re;
}
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

So you are trying to get multiple kernels in a single kernel? For a single equation or for multiple terms in a single equation?
Either way this is at a lower level than we'd advise users to work at. If you could please work with regular kernels first it ll be easier to support you. 

If not you may use the MOOSE doxygen to get more information about all these routines
https://mooseframework.inl.gov/docs/doxygen/moose/classTaggingInterface.html#a7db1f9ea23c3f5fb0473118984dc4c0f
Guillaume

**Comment by YaqiWang** [comment]
I thought you decided to use ArrayKernel.
